java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -242,21 +246,38 @@ public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
 
     @Override
     public void visitBinaryExpression(BinaryExpressionTree tree) {
-      if(tree.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
-        increaseComplexityByOne(tree.operatorToken());
-        ExpressionTree left = ExpressionUtils.skipParentheses(tree.leftOperand());
-        if(left.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) left).operatorToken());
-          complexity--;
-        }
-        ExpressionTree right = ExpressionUtils.skipParentheses(tree.rightOperand());
-        if(right.kind() == tree.kind()) {
-          ignored.add(((BinaryExpressionTree) right).operatorToken());
-          complexity--;
+      if (tree.is(CONDITIONAL_AND, CONDITIONAL_OR) && !ignored.contains(tree)) {
+        List<BinaryExpressionTree> flatOperators = flattenLogicalExpression(tree);
+
+        // first binary expression
+        increaseComplexityByOne(flatOperators.get(0).operatorToken());
+
+        // rest
+        for (int i = 1; i < flatOperators.size(); i++) {
```

Comment:
matter of style but this could be rewritten like : 
```
 BinaryExpressionTree previous = null;
for( BinaryExpressionTree current : flatOperators) { 
         if (previous == null || !previous.is(current.kind())) {
             increaseComplexityByOne(current.operatorToken());
         }
        previous = current;
}
```